### PR TITLE
Delete completed template job so no need to delete it when rerunning the job

### PR DESCRIPTION
**Pull request by malayparida2000**
Resolves-https://bugzilla.redhat.com/show_bug.cgi?id=1970939

This eliminates the need to delete the older/existing job when
 triggering the job for a second time.

### CODE REVIEW

**Review by travisn — COMMENTED** [review]
/lgtm

File: controllers/storagecluster/job_templates.go [hunk 1]
```diff
@@ -57,6 +58,21 @@ func (obj *ocsJobTemplates) ensureCreated(r *StorageClusterReconciler, sc *ocsv1
...
+		// Delete a template job after the job gets completed
+		// This eliminates the need to delete the older/existing job when triggering the job for a second time
+		job := &batchv1.Job{}
+		err = r.Client.Get(r.ctx, client.ObjectKey{Name: fmt.Sprintf("%s-job", template.Name), Namespace: template.Namespace}, job)
+		if err == nil {
+			if job.Status.Succeeded >= 1 {
+				r.Log.Info(fmt.Sprintf("Job %s is completed", job.Name))
+				err = r.Client.Delete(r.ctx, job)
+				if err != nil {
+					return reconcile.Result{}, fmt.Errorf("failed to delete job %s: %v", template.Name, err.Error())
```
Comment (umangachapagain) [inline controllers/storagecluster/job_templates.go:71]: If we can't delete the job, should we block rest of the reconcile?
IMO we should try to delete but not return an error.

File: controllers/storagecluster/job_templates.go [hunk 2]
```diff
@@ -57,6 +58,21 @@ func (obj *ocsJobTemplates) ensureCreated(r *StorageClusterReconciler, sc *ocsv1
...
 			return reconcile.Result{}, fmt.Errorf("failed to create Template : %v", err.Error())
 		}
+
+		// Delete a template job after the job gets completed
+		// This eliminates the need to delete the older/existing job when triggering the job for a second time
+		job := &batchv1.Job{}
+		err = r.Client.Get(r.ctx, client.ObjectKey{Name: fmt.Sprintf("%s-job", template.Name), Namespace: template.Namespace}, job)
+		if err == nil {
+			if job.Status.Succeeded >= 1 {
+				r.Log.Info(fmt.Sprintf("Job %s is completed", job.Name))
```
Comment (umangachapagain) [inline controllers/storagecluster/job_templates.go:68]: This isn't a very helpful message. Let's drop it and log it after we delete it.

File: controllers/storagecluster/job_templates.go [hunk 3]
```diff
@@ -57,6 +58,21 @@ func (obj *ocsJobTemplates) ensureCreated(r *StorageClusterReconciler, sc *ocsv1
...
+				err = r.Client.Delete(r.ctx, job)
+				if err != nil {
+					return reconcile.Result{}, fmt.Errorf("failed to delete job %s: %v", template.Name, err.Error())
+				}
+				r.Log.Info(fmt.Sprintf("Job %s is deleted after completion", job.Name))
```
Comment (umangachapagain) [inline controllers/storagecluster/job_templates.go:73]: Needs more detail. What is this job related to?
Also it's better to avoid `sprintf` and use structured logging format as used in other places.

Comment (malayparida2000) [reply controllers/storagecluster/job_templates.go]: We have job templates, Which can be invoked with oc process command to create jobs to complete some tasks.
The job names will give the context itself. We have two templates ocs-osd-removal & ocs-extend-cluster.

This is how the logs will look like
```
~ $ oc logs ocs-operator-8676bdf8bb-hgr9q | grep "deleted successfully"
{"level":"info","ts":"2023-08-10T10:30:09Z","logger":"controllers.StorageCluster","msg":"Job ocs-osd-removal-job deleted successfully after completion","Request.Namespace":"openshift-storage","Request.Name":"ocs-storagecluster"}
```

File: controllers/storagecluster/job_templates.go [hunk 4]
```diff
@@ -57,6 +58,20 @@ func (obj *ocsJobTemplates) ensureCreated(r *StorageClusterReconciler, sc *ocsv1
 		if err != nil && !errors.IsAlreadyExists(err) {
 			return reconcile.Result{}, fmt.Errorf("failed to create Template : %v", err.Error())
 		}
+
+		// Delete a template job after the job gets completed
```
Comment (Madhu-1) [inline controllers/storagecluster/job_templates.go:62]: the order of sequence should be like below right?

```
job:=Get()
if err!=nil && !NotPresent{
return nil,err
}
if error==nil && job.Status.Succeeded >= 1{
// delete it and return
} 
err=CreateOrUpdate()
return err
```

Comment (malayparida2000) [reply controllers/storagecluster/job_templates.go:62]: Hi Madhu, actually no. While creating we create the openshift templatest only. When a user wants they create the job by using the oc process command. 

```
oc process -n openshift-storage ocs-osd-removal -p FAILED_OSD_IDS=0 FORCE_OSD_REMOVAL=true |oc create -n openshift-storage -f -
```

Comment (Madhu-1) [reply controllers/storagecluster/job_templates.go:62]: Got it, Thank you

File: controllers/storagecluster/job_templates.go [hunk 5]
```diff
@@ -57,6 +58,20 @@ func (obj *ocsJobTemplates) ensureCreated(r *StorageClusterReconciler, sc *ocsv1
...
+		// This eliminates the need to delete the older/existing job when triggering the job for a second time
+		job := &batchv1.Job{}
+		err = r.Client.Get(r.ctx, client.ObjectKey{Name: fmt.Sprintf("%s-job", template.Name), Namespace: template.Namespace}, job)
+		if err == nil {
+			if job.Status.Succeeded >= 1 {
+				err = r.Client.Delete(r.ctx, job)
+				if err != nil {
+					r.Log.Info("Job " + job.Name + " couldn't be deleted successfully after completion")
+				}
```
Comment (travisn) [inline controllers/storagecluster/job_templates.go:71]: Line 72 will be mistakenly logged if there is an error
```suggestion
				} else {
```

**Review by travisn — CHANGES_REQUESTED** [review]

File: controllers/storagecluster/job_templates.go [hunk 6]
```diff
@@ -57,6 +58,20 @@ func (obj *ocsJobTemplates) ensureCreated(r *StorageClusterReconciler, sc *ocsv1
...
+
+		// Delete a template job after the job gets completed
+		// This eliminates the need to delete the older/existing job when triggering the job for a second time
+		job := &batchv1.Job{}
+		err = r.Client.Get(r.ctx, client.ObjectKey{Name: fmt.Sprintf("%s-job", template.Name), Namespace: template.Namespace}, job)
+		if err == nil {
+			if job.Status.Succeeded >= 1 {
+				err = r.Client.Delete(r.ctx, job)
+				if err != nil {
+					r.Log.Info("Job " + job.Name + " couldn't be deleted successfully after completion")
```
Comment (umangachapagain) [inline controllers/storagecluster/job_templates.go:70]:
```suggestion
					r.Log.Info("Job couldn't be deleted successfully after completion.", "Job", klog.KRef(job.Namespace, job.Name))
```

File: controllers/storagecluster/job_templates.go [hunk 7]
```diff
@@ -57,6 +58,20 @@ func (obj *ocsJobTemplates) ensureCreated(r *StorageClusterReconciler, sc *ocsv1
...
+				}
+				r.Log.Info("Job " + job.Name + " deleted successfully after completion")
```
Comment (umangachapagain) [inline controllers/storagecluster/job_templates.go:72]:
```suggestion
				r.Log.Info("Job deleted successfully after completion.", "Job", klog.KRef(job.Namespace, job.Name))
```

**Review by umangachapagain — CHANGES_REQUESTED** [review]

File: controllers/storagecluster/job_templates.go [hunk 8]
```diff
@@ -57,6 +59,21 @@ func (obj *ocsJobTemplates) ensureCreated(r *StorageClusterReconciler, sc *ocsv1
...
 			return reconcile.Result{}, fmt.Errorf("failed to create Template : %v", err.Error())
 		}
+
+		// Delete a template job after the job gets completed
+		// This eliminates the need to delete the older/existing job when triggering the job for a second time
+		job := &batchv1.Job{}
+		err = r.Client.Get(r.ctx, client.ObjectKey{Name: fmt.Sprintf("%s-job", template.Name), Namespace: template.Namespace}, job)
+		if err == nil {
+			if job.Status.Succeeded >= 1 {
+				err = r.Client.Delete(r.ctx, job)
```
[the remark this answers is not in the document]
Comment (malayparida2000) [reply controllers/storagecluster/job_templates.go:69]: Yes,
https://github.com/red-hat-storage/ocs-operator/blob/58c0a5630cc2511a02f533516bc94cc85ee3fc92/config/rbac/role.yaml#L29-L38

https://github.com/red-hat-storage/ocs-operator/blob/58c0a5630cc2511a02f533516bc94cc85ee3fc92/controllers/storagecluster/reconcile.go#L128

**Review by umangachapagain — APPROVED** [review]
